File: ts/wasm/lib/detailed/chain_api.ts
import type { IHiveChainInterface, IManabarData, ITransaction, IOnlineTransaction, TTimestamp, TPublicKey, TWaxExtended, TBlockHash, TWaxRestExtended, TDeepWaxApiRequestPartial, IWaxOptionsChain, TNaiAssetConvertible } from "./interfaces";
import type { MainModule, MapStringUInt16, wax_authority, wax_authorities } from "../build_wasm/wax.common";
import { ApiAuthority, ApiWitness, type ApiAccount, type ApiManabar, type ApiTransaction, type RcAccount } from "./api";

import { WaxError, WaxChainApiError } from "./errors.js";
import { ONE_HUNDRED_PERCENT, WaxBaseApi } from "./base_api.js";
import { HiveApiTypes, HiveRestApiTypes } from "./chain_api_data.js";
import { iterate } from "./util/iterate.js";
import { OnlineTransaction} from "./online_transaction";
import { dateFromString } from "./util/expiration_parser.js";

import { ApiCaller, TRequestInterceptor, TResponseInterceptor } from "./util/api_caller";

import { TAccountName } from "./hive_apps_operations";
import { ISignatureProvider } from "./extensions/signatures";
import { WasmManager } from "./util/wasm_errors";

export enum EManabarType {
  UPVOTE = 0,
  DOWNVOTE = 1,
  RC = 2
}

export type TChainReferenceData = {
  head_block_id: TBlockHash,
  head_block_time: Date
};

export enum EChainApiType {
  JSON_RPC = "json_rpc",
  REST = "rest"
}

export type TAccountAuthorityCollection = Map<TAccountName, [wax_authorities, TPublicKey]>;

export class HiveChainApi extends WaxBaseApi implements IHiveChainInterface {
  public get restApi () {
    return this.restApiCaller.createApiCaller() as unknown as IHiveChainInterface['restApi'];
  }

  public get api () {
    return this.jsonRpcApiCaller.createApiCaller() as unknown as IHiveChainInterface['api'];
  }

  private jsonRpcApiCaller: ApiCaller;
  private restApiCaller: ApiCaller;

  private taposCache: TChainReferenceData = { head_block_id: '', head_block_time: new Date(Date.now()) };
  private lastTaposCacheUpdate: number = 0; /// last timestamp of taposCache update (in milliseconds)

  private readonly apiTimeout: number;

  public constructor(
    wasmManager: WasmManager,
    wax: MainModule,
    config: IWaxOptionsChain,
    public readonly originator: HiveChainApi|null
  ) {
    super(wasmManager, wax, config.chainId);

    this.apiTimeout = config.apiTimeout;

    this.jsonRpcApiCaller = new ApiCaller(EChainApiType.JSON_RPC, config.apiEndpoint, this.apiTimeout, iterate(originator ? structuredClone(originator.jsonRpcApiCaller.localTypes) : {}, HiveApiTypes), 'POST', (path, newValue, found) => {
      if (this.originator !== null) // Propagate the change to the originator
        return found ||= this.originator.jsonRpcApiCaller.setEndpointUrlForPath(path, newValue, found);

      return found || false;
    }, config.waxApiCaller, data => { // Rewrite request data to JSON-RPC format
        data.data = JSON.stringify({
        jsonrpc: "2.0",
        method: data.paths.join('.'),
        params: data.data,
        id: 1
      });
      data.url = '';

      return data;
    }, data => { // Rewrite response data to JSON-RPC format
      if (typeof data.response === "object") {
        if ("result" in data.response) {
          data.response = data.response.result;
          return data;
        }

        if ("error" in data.response && typeof data.response.error === "object" && "data" in data.response.error) {
          // Possibly an exception that we can recognize & repackage.
          this.wasmManager.safeWasmCall(() => this.protocol.cpp_transform_api_error_response_into_exception(JSON.stringify(data.response.error.data)));
        }
      }

      throw new WaxChainApiError(`Invalid response from chain API`, data.response);
    });

    this.restApiCaller = new ApiCaller(EChainApiType.REST, config.restApiEndpoint, this.apiTimeout, iterate(originator ? structuredClone(originator.restApiCaller.localTypes) : {}, HiveRestApiTypes), 'GET', (path, newValue, found) => {
      if (this.originator !== null) // Propagate the change to the originator
        return found ||= this.originator.restApiCaller.setEndpointUrlForPath(path, newValue, found);

      return found || false;
    }, config.waxApiCaller);
  }

  public async broadcast(transaction: ApiTransaction | ITransaction | IOnlineTransaction): Promise<void> {
    const toBroadcast: object = "toApiJson" in transaction ? transaction.toApiJson() : transaction;

    if ("performOnChainVerification" in transaction)
      await transaction.performOnChainVerification();

    await this.api.network_broadcast_api.broadcast_transaction({
      max_block_age: -1,
      trx: toBroadcast as ApiTransaction
    });
  }

  public withProxy(requestInterceptor: TRequestInterceptor, responseInterceptor: TResponseInterceptor): HiveChainApi {
    const newInstance = this.extend();
    newInstance.jsonRpcApiCaller.requestInterceptor = requestInterceptor;
    newInstance.jsonRpcApiCaller.responseInterceptor = responseInterceptor;
    newInstance.restApiCaller.requestInterceptor = requestInterceptor;
    newInstance.restApiCaller.responseInterceptor = responseInterceptor;

    return newInstance;
  }

  public set endpointUrl(endpoint: string) {
    this.jsonRpcApiCaller.defaultEndpointUrl = endpoint;

    if(this.originator !== null)
      this.originator.jsonRpcApiCaller.defaultEndpointUrl = endpoint;
  }

  public get endpointUrl(): string {
    return this.jsonRpcApiCaller.defaultEndpointUrl;
  }

  public extend<YourApi>(extendedHiveApiData?: YourApi): HiveChainApi & TWaxExtended<YourApi, this> {
    const newApi = new HiveChainApi(this.wasmManager, this.wax, {
      chainId: this.chainId,
      apiEndpoint: this.jsonRpcApiCaller.defaultEndpointUrl,
      restApiEndpoint: this.restApiCaller.defaultEndpointUrl,
      apiTimeout: this.apiTimeout,
      waxApiCaller: this.jsonRpcApiCaller.defaultWaxApiCaller,
    }, this);

    if(typeof extendedHiveApiData === "object")
      iterate(newApi.jsonRpcApiCaller.localTypes, extendedHiveApiData as object)

    return newApi as unknown as HiveChainApi & TWaxExtended<YourApi, this>;
  }

  public extendRest<YourRestApi>(extendedHiveRestApiData?: TDeepWaxApiRequestPartial<YourRestApi>): HiveChainApi & TWaxRestExtended<YourRestApi, this> {
    const newApi = new HiveChainApi(this.wasmManager, this.wax, {
      chainId: this.chainId,
      apiEndpoint: this.jsonRpcApiCaller.defaultEndpointUrl,
      restApiEndpoint: this.restApiCaller.defaultEndpointUrl,
      apiTimeout: this.apiTimeout,
      waxApiCaller: this.jsonRpcApiCaller.defaultWaxApiCaller,
    }, this);

    if(typeof extendedHiveRestApiData === "object")
      iterate(newApi.restApiCaller.localTypes, extendedHiveRestApiData as object);

    return newApi as unknown as HiveChainApi & TWaxRestExtended<YourRestApi, this>;
  }

  public async createTransaction(expirationTime?: TTimestamp): Promise<IOnlineTransaction> {
    const chainReferenceData = await this.acquireChainReferenceData(3000);

    const transaction = new OnlineTransaction(this, chainReferenceData, expirationTime); 
    return transaction;
  }

  private async acquireChainReferenceData(taposLiveness: number): Promise<TChainReferenceData> {
    const now = Date.now();
    if ((now - this.lastTaposCacheUpdate) >= taposLiveness) {
      const { head_block_id, time } = await this.api.database_api.get_dynamic_global_properties({});
      this.taposCache = { head_block_id: head_block_id, head_block_time: dateFromString(time) };
      this.lastTaposCacheUpdate = now;
    }

    return this.taposCache;
  }

  private async getRcManabarForAccount(accountName: string): Promise<RcAccount> {
    const { rc_accounts: [ account ] } = await this.api.rc_api.find_rc_accounts({ accounts: [ accountName ] });
    if(typeof account === "undefined")
      throw new WaxError(`No such account on chain with given name: "${accountName}"`);

    return account;
  }

  private async findAccountsNoThrow(...accountNames: string[]): Promise<Array<ApiAccount>> {
    const { accounts } = await this.api.database_api.find_accounts({ accounts: accountNames, delayed_votes_active: true });
    return accounts;
  }

  private async findAccounts(...accountNames: string[]): Promise<Array<ApiAccount>> {
    const accounts = await this.findAccountsNoThrow(...accountNames);
    if(accounts.length !== accountNames.length) {
      const notFoundAccounts = accounts.map(node => node.name).filter(node => !accountNames.includes(node));

      throw new WaxError(`No such account(s) on chain with given name(s): "${notFoundAccounts.join(', ')}"`);
    }

    return accounts;
  }

  private async findWitnessAccountsNoThrow(...witnessNames: string[]): Promise<Array<ApiWitness>> {
    const { witnesses } = await this.api.database_api.find_witnesses({ owners: witnessNames, delayed_votes_active: true });
    return witnesses;
  }

  private async findWitnessAccounts(...witnessNames: string[]): Promise<Array<ApiWitness>> {
    const accounts = await this.findWitnessAccountsNoThrow(...witnessNames);
    if(accounts.length !== witnessNames.length) {
      const notFoundAccounts = accounts.map(node => node.owner).filter(node => !witnessNames.includes(node));

      throw new WaxError(`No such witness(s) on chain with given name(s): "${notFoundAccounts.join(', ')}"`);
    }

    return accounts;
  }

  private async findAccount(accountName: string): Promise<ApiAccount> {
    const [ account ] = await this.findAccounts(accountName);

    return account;
  }

  public transformApiAuthority(input: ApiAuthority): wax_authority {
    const transformEntries = (input: {"0": string; "1": number}[], storage: MapStringUInt16) => input.reduce((storage: MapStringUInt16, data): MapStringUInt16 => {
      storage.set(data[0], data[1]);
      return storage;
      }, storage);

      const retVal: wax_authority = {
        weight_threshold: input.weight_threshold,
        account_auths: transformEntries(input.account_auths, new this.wax.MapStringUInt16()),
        key_auths: transformEntries(input.key_auths, new this.wax.MapStringUInt16())
      };

      return retVal;
  }

  public async collectAccountAuthorities(throwIfMissing: boolean, ...accountNames: string[]): Promise<TAccountAuthorityCollection> {
    if(accountNames.length === 0)
      return new Map();

    const accountData = throwIfMissing ? await this.findAccounts(...accountNames) : await this.findAccountsNoThrow(...accountNames);

    const retVal = new Map<TAccountName, [wax_authorities, TPublicKey]>();

    for(let i = 0; i < accountData.length; ++i) {
      const name = accountData[i].name;
      const owner = accountData[i].owner;
      const active = accountData[i].active;
      const posting = accountData[i].posting;

      const account_authority: wax_authorities = {
        owner: this.transformApiAuthority(owner),
        active: this.transformApiAuthority(active),
        posting: this.transformApiAuthority(posting)
      };

      retVal.set(name, [account_authority, accountData[i].memo_key]);
    }

    return retVal;
  }

  public async collectWitnessSigningKeys(throwIfMissing: boolean, ...accountNames: string[]): Promise<Map<TAccountName, TPublicKey>> { 
    if(accountNames.length === 0)
      return new Map();

    const accountData = throwIfMissing ? await this.findWitnessAccounts(...accountNames) : await this.findWitnessAccountsNoThrow(...accountNames);

    const retVal = new Map<TAccountName, TPublicKey>();

    for(let i = 0; i < accountData.length; ++i)
      retVal.set(accountData[i].owner, accountData[i].signing_key);

    return retVal;
  }

  public async encryptForAccounts(wallet: ISignatureProvider, content: string, fromAccount: string, toAccount: string): Promise<string> {
    let from: string, to: string;

    ([ { memo_key: from }, { memo_key: to } ] = await this.findAccounts(...[... new Set([ fromAccount, toAccount ])]));

    if(to === undefined)
      to = from;

    const encrypted = wallet.encryptData(content, from, to);

    return this.wasmManager.safeWasmCall(() => this.protocol.cpp_crypto_memo_dump_string({
      content: encrypted,
      from,
      to
    }));
  }

  private async getManabarDataArguments(accountName: string, manabarType: EManabarType): Promise<Parameters<WaxBaseApi['calculateCurrentManabarValue']>> {
    const dgpo = await this.api.database_api.get_dynamic_global_properties({});

    let manabar: ApiManabar;
    let max: TNaiAssetConvertible;

    if(manabarType === EManabarType.RC) {
      ({ rc_manabar: manabar, max_rc: max } = await this.getRcManabarForAccount(accountName));
    } else {
      const account = await this.findAccount(accountName);

      manabar = manabarType === EManabarType.UPVOTE ? account.voting_manabar : account.downvote_manabar;
      max = BigInt(account.post_voting_power.amount);

      if(manabarType === EManabarType.DOWNVOTE) {
        const downvotePoolPercent = BigInt(dgpo.downvote_pool_percent);
        if(max / ONE_HUNDRED_PERCENT > ONE_HUNDRED_PERCENT)
          max = (max / ONE_HUNDRED_PERCENT) * downvotePoolPercent;
        else
          max = (max * downvotePoolPercent) / ONE_HUNDRED_PERCENT;
      }
    }

    return [
      Math.round(new Date(`${dgpo.time}Z`).getTime() / 1000), // Convert API time to seconds
      max,
      manabar.current_mana,
      manabar.last_update_time
    ];
  }

  public async calculateCurrentManabarValueForAccount(accountName: string, manabarType: EManabarType = EManabarType.UPVOTE): Promise<IManabarData> {
    const args = await this.getManabarDataArguments(accountName, manabarType);

    return super.calculateCurrentManabarValue(
      ...args
    );
  }

  public async calculateManabarFullRegenerationTimeForAccount(accountName: string, manabarType: EManabarType = EManabarType.UPVOTE): Promise<Date> {
    const args = await this.getManabarDataArguments(accountName, manabarType);

    if (args[1] === 0n)
      return new Date();

    const time = super.calculateManabarFullRegenerationTime(
      ...args
    );

    return new Date(time * 1000);
  }
}

File: python/wax/_private/chain_api.py
from __future__ import annotations

import time
from typing import TYPE_CHECKING, Final, Generic, cast

from wax._private.api.api_caller import WaxApiCaller
from wax._private.api.models import ApiTransaction
from wax._private.base_api import WaxBaseApi
from wax._private.converters.url_converter import convert_to_http_url
from wax._private.models.hive_date_time import HiveDateTime
from wax._private.online_transaction import OnlineTransaction
from wax.api.collection import WaxApiCollection
from wax.exceptions.chain_errors import AccountNotFoundError
from wax.exceptions.validation_errors import InvalidAccountNameError
from wax.interfaces import ApiCollectionT, ExtendedApiCollectionT, IHiveChainInterface, IOnlineTransaction
from wax.models.authority import WaxAccountAuthorityInfo, WaxAuthorities, WaxAuthority
from wax.models.basic import ChainReferenceData

if TYPE_CHECKING:
    from datetime import datetime, timedelta

    from beekeepy.interfaces import HttpUrl
    from wax._private.api.models import FindAccountsApiResponse, PossibleAuthorityApi
    from wax.interfaces import ITransaction
    from wax.models.basic import AccountName, ChainId


class HiveChainApi(IHiveChainInterface, WaxBaseApi, Generic[ApiCollectionT]):
    TAPOS_LIVENESS: Final[int] = 3000  # 3 seconds / 3000 milliseconds

    def __init__(
        self,
        chain_id: ChainId,
        endpoint_url: HttpUrl,
        api_collection: type[ApiCollectionT] | None = None,
        *,
        _private: bool = False,
    ) -> None:
        """
        Initializes HiveChainApi.

        Args:
            chain_id: chain id of the node passed by `node_address`.
            endpoint_url: address of the node to connect to.
            api_collection: type of the api collection to use. Defaults to WaxApiCollection.
            _private: flag indicating that the constructor should not be called directly.

        Raises:
            AssertionError: if the constructor is called directly.
        """
        assert _private, "HiveChainApi should not be instantiated directly. Use create_hive_chain() instead."
        super().__init__(chain_id, _private=True)
        self._chain_id = chain_id
        self._endpoint_url = endpoint_url

        if api_collection is None:
            api_collection = WaxApiCollection  # type: ignore[assignment]  # ApiCollectionT can be WaxApiCollection

        assert api_collection is not None, "api_collection must be provided or default to WaxApiCollection."
        self._api_collection = api_collection
        self._api_caller = WaxApiCaller(api_collection(), endpoint_url)

        self._last_tapos_cache_update = 0
        self._tapos_cache = ChainReferenceData(time=HiveDateTime.now())

    @property
    def api(self) -> ApiCollectionT:  # type: ignore[override]
        return self._api_caller.api

    def extends(self, new_api: type[ExtendedApiCollectionT]) -> HiveChainApi[ExtendedApiCollectionT | ApiCollectionT]:  # type: ignore[override]
        current_api: type[ApiCollectionT | ExtendedApiCollectionT] = self._api_collection

        class NewApi(new_api, current_api):  # type: ignore[valid-type, misc]
            def __init__(self) -> None:
                new_api.__init__(self)
                current_api.__init__(self)
                # super().__init__() is not used here as it requires the user to call the super() method also,
                # which causes that it is required when some collection has not base class

        return HiveChainApi[ExtendedApiCollectionT | ApiCollectionT](
            self.chain_id, self.endpoint_url, NewApi, _private=True
        )

    @property
    def endpoint_url(self) -> HttpUrl:
        return self._endpoint_url

    @endpoint_url.setter
    def endpoint_url(self, value: HttpUrl | str) -> None:
        parsed_endpoint_url = convert_to_http_url(value)

        self._endpoint_url = parsed_endpoint_url
        self._api_caller.set_endpoint_url(parsed_endpoint_url, _private=True)

    def teardown(self) -> None:
        """Call when work with communicator is over."""
        self._api_caller.teardown()

    async def create_transaction(self, expiration: datetime | timedelta | None = None) -> IOnlineTransaction:
        chain_reference_data = await self._acquire_chain_reference_data()
        expiration = self._resolve_expiration(expiration)
        return OnlineTransaction(self, chain_reference_data, expiration)  # type: ignore[arg-type]

    async def broadcast(self, transaction: ITransaction | IOnlineTransaction) -> None:
        if isinstance(transaction, IOnlineTransaction):
            await transaction.perform_on_chain_verification()

        await self._internal_api.network_broadcast_api.broadcast_transaction(
            trx=ApiTransaction(**transaction.to_dict()), max_block_age=-1
        )

    async def collect_account_authorities(
        self, *accounts: AccountName
    ) -> WaxAccountAuthorityInfo | list[WaxAccountAuthorityInfo]:
        for account in accounts:
            if not self.is_valid_account_name(account):
                raise InvalidAccountNameError(account)

        accounts_response = await self._internal_api.database_api.find_accounts(accounts=[*accounts])

        if len(accounts) == 1:
            return self._extract_authority_from_find_accounts_response(accounts_response, accounts[0])[0]

        return self._extract_authority_from_find_accounts_response(accounts_response, *accounts)

    def extend_rest(  # type: ignore[override]
        self, new_rest_api: type[ExtendedApiCollectionT]
    ) -> HiveChainApi[ExtendedApiCollectionT | ApiCollectionT]:
        return self.extends(new_rest_api)

    @property
    def _internal_api(self) -> WaxApiCollection:
        return cast(WaxApiCollection, self._api_caller.api)

    async def _acquire_chain_reference_data(self) -> ChainReferenceData:
        now = self._current_milli_time()

        if now - self._last_tapos_cache_update > self.TAPOS_LIVENESS:
            dgpo = await self._internal_api.database_api.get_dynamic_global_properties()
            self._tapos_cache = ChainReferenceData(
                head_block_id=dgpo.head_block_id,
                time=HiveDateTime(dgpo.time),
            )
            self._last_tapos_cache_update = now

        return self._tapos_cache

    def _transform_api_authority(self, api_authority: PossibleAuthorityApi) -> WaxAuthority | None:
        if not api_authority:
            return None

        entity_index: Final[int] = 0
        weight_index: Final[int] = 1

        return WaxAuthority(
            weight_threshold=api_authority.weight_threshold,
            account_auths={account[entity_index]: account[weight_index] for account in api_authority.account_auths},
            key_auths={key[entity_index]: cast(int, key[weight_index]) for key in api_authority.key_auths},
        )

    def _extract_authority_from_find_accounts_response(
        self,
        response: FindAccountsApiResponse,
        *searched_accounts: AccountName,
    ) -> list[WaxAccountAuthorityInfo]:
        if not response.accounts:
            raise AccountNotFoundError(*searched_accounts)

        accounts_to_search = set(searched_accounts)
        accounts_data = []

        for account in response.accounts:
            if account.name in searched_accounts:
                accounts_data.append(account)
                accounts_to_search.remove(account.name)

                if not accounts_to_search:
                    break

        if not accounts_data:
            raise AccountNotFoundError(*accounts_to_search)

        return [
            WaxAccountAuthorityInfo(
                account_data.name,
                WaxAuthorities(
                    owner=self._transform_api_authority(account_data.owner),
                    active=self._transform_api_authority(account_data.active),
                    posting=self._transform_api_authority(account_data.posting),
                ),
                account_data.memo_key,
                HiveDateTime(account_data.last_owner_update),
                HiveDateTime(account_data.previous_owner_update),
            )
            for account_data in accounts_data
        ]

    def _current_milli_time(self) -> int:
        return round(time.time() * 1000)
